src/dbnode/storage/namespace.go

```diff
@@ -924,20 +926,90 @@ func (n *dbNamespace) Flush(
 		}
 
 		// skip flushing if the shard has already flushed data for the `blockStart`
-		if s := shard.FlushState(blockStart); s.Status == fileOpSuccess {
+		if s := shard.FlushState(blockStart); s.WarmStatus == fileOpSuccess {
 			continue
 		}
 		// NB(xichen): we still want to proceed if a shard fails to flush its data.
 		// Probably want to emit a counter here, but for now just log it.
-		if err := shard.Flush(blockStart, flushPersist); err != nil {
+		if err := shard.WarmFlush(blockStart, flushPersist); err != nil {
 			detailedErr := fmt.Errorf("shard %d failed to flush data: %v",
 				shard.ID(), err)
 			multiErr = multiErr.Add(detailedErr)
 		}
 	}
 
 	res := multiErr.FinalError()
-	n.metrics.flush.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	n.metrics.flushWarmData.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
+
+// idAndBlockStart is the composite key for the genny map used to keep track of
+// dirty series that need to be ColdFlushed.
+type idAndBlockStart struct {
+	id         ident.ID
+	blockStart xtime.UnixNano
+}
+
+type coldFlushReuseableResources struct {
+	// dirtySeries is a map from a composite key of <series ID, block start>
+	// to an element in a list in the dirtySeriesToWrite map. This composite key
+	// is deliberately made so that this map stays one level deep, making it
+	// easier to share between shard loops, minimizing the need for allocations.
+	dirtySeries *dirtySeriesMap
+	// dirtySeriesToWrite is a map from block start to a list of dirty series
+	// that have yet to be written to disk.
+	dirtySeriesToWrite map[xtime.UnixNano]*idList
+	// idElementPool is a pool of list elements to be used when constructing
+	// new lists for the dirtySeriesToWrite map.
+	idElementPool *idElementPool
+	fsReader      fs.DataFileSetReader
+}
+
+func (n *dbNamespace) ColdFlush(
+	flushPersist persist.FlushPreparer,
+) error {
+	// NB(rartoul): This value can be used for emitting metrics, but should not be used
+	// for business logic.
+	callStart := n.nowFn()
+
+	n.RLock()
+	if n.bootstrapState != Bootstrapped {
+		n.RUnlock()
+		n.metrics.flushColdData.ReportError(n.nowFn().Sub(callStart))
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.ColdWritesEnabled() {
+		n.metrics.flushColdData.ReportSuccess(n.nowFn().Sub(callStart))
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.GetOwnedShards()
+
+	fsReader, err := fs.NewReader(n.opts.BytesPool(), n.opts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return err
+	}
+	resources := coldFlushReuseableResources{
+		dirtySeries:        newDirtySeriesMap(dirtySeriesMapOptions{}),
```

Comment:
I think it's good not to reuse these between cold flushes since it means we don't bloat process at all for most workloads since they shouldn't do it often/at all and it wouldn't any extra overhead for those.

For the case where it happens often, they know already they're not doing the ideal thing (writing into the past often) and can likely deal with a bit of extra GC.